File: plugins-sdk/src/api/dialog.ts
import { invoke } from '../core/ipc';
import { dispatch } from '../core/dispatch';

// Dialog 错误类型
export interface DialogError extends Error {
  name: 'DialogError';
  code?: string;
}

export function createDialogError(message: string, code?: string): DialogError {
  const error = new Error(message) as DialogError;
  error.name = 'DialogError';
  error.code = code;
  return error;
}

export function isDialogError(error: any): error is DialogError {
  return error && error.name === 'DialogError';
}

// 消息对话框选项
export interface MessageDialogOptions {
  title?: string;
  message: string;
  kind?: 'info' | 'warning' | 'error';
  okLabel?: string;
}

// 确认对话框选项
export interface ConfirmDialogOptions {
  title?: string;
  message: string;
  kind?: 'info' | 'warning' | 'error';
  okLabel?: string;
  cancelLabel?: string;
}

// 文件对话框过滤器
export interface DialogFilter {
  name: string;
  extensions: string[];
}

// 打开文件对话框选项
export interface OpenDialogOptions {
  title?: string;
  defaultPath?: string;
  filters?: DialogFilter[];
  multiple?: boolean;
  directory?: boolean;
}

// 保存文件对话框选项
export interface SaveDialogOptions {
  title?: string;
  defaultPath?: string;
  filters?: DialogFilter[];
}

// 通用的对话框调用辅助函数
function callDialogApi<T = any>(method: string, args?: any): Promise<T> {
  return dispatch({
    webview: () => invoke<T>(method, args),
    headless: () => invoke<T>(method, args),
  });
}

/**
 * 显示消息对话框
 * @param options 消息对话框选项
 */
export function showMessage(options: MessageDialogOptions): Promise<void> {
  return callDialogApi("plugin_dialog_message", options);
}

/**
 * 显示确认对话框
 * @param options 确认对话框选项
 * @returns 用户是否点击了确认按钮
 */
export function showConfirm(options: ConfirmDialogOptions): Promise<boolean> {
  return callDialogApi<boolean>("plugin_dialog_confirm", options);
}

/**
 * 显示打开文件对话框
 * @param options 打开文件对话框选项
 * @returns 选择的文件路径，如果取消则返回 null
 */
export async function showOpen(options?: OpenDialogOptions): Promise<string | string[] | null> {
  const result = await callDialogApi<any>("plugin_dialog_open", options || {});
  
  // 处理返回值类型转换
  if (result === null || result === undefined) {
    return null;
  }
  
  // 如果是数组，说明是多文件选择
  if (Array.isArray(result)) {
    return result as string[];
  }
  
  // 如果是字符串，说明是单文件选择
  if (typeof result === 'string') {
    return result;
  }
  
  return null;
}

/**
 * 显示保存文件对话框
 * @param options 保存文件对话框选项
 * @returns 选择的保存路径，如果取消则返回 null
 */
export function showSave(options?: SaveDialogOptions): Promise<string | null> {
  return callDialogApi<string | null>("plugin_dialog_save", options || {});
}

// 便捷方法
/**
 * 显示信息消息
 * @param message 消息内容
 * @param title 标题（可选）
 */
export function info(message: string, title?: string): Promise<void> {
  return showMessage({
    message,
    title,
    kind: 'info'
  });
}

/**
 * 显示警告消息
 * @param message 消息内容
 * @param title 标题（可选）
 */
export function warning(message: string, title?: string): Promise<void> {
  return showMessage({
    message,
    title,
    kind: 'warning'
  });
}

/**
 * 显示错误消息
 * @param message 消息内容
 * @param title 标题（可选）
 */
export function error(message: string, title?: string): Promise<void> {
  return showMessage({
    message,
    title,
    kind: 'error'
  });
}

/**
 * 显示确认对话框（简化版）
 * @param message 消息内容
 * @param title 标题（可选）
 * @returns 用户是否点击了确认按钮
 */
export function confirm(message: string, title?: string): Promise<boolean> {
  return showConfirm({
    message,
    title
  });
}

/**
 * 选择单个文件
 * @param filters 文件过滤器（可选）
 * @param defaultPath 默认路径（可选）
 * @returns 选择的文件路径，如果取消则返回 null
 */
export function selectFile(filters?: DialogFilter[], defaultPath?: string): Promise<string | null> {
  return showOpen({
    filters,
    defaultPath,
    multiple: false,
    directory: false
  }) as Promise<string | null>;
}

/**
 * 选择多个文件
 * @param filters 文件过滤器（可选）
 * @param defaultPath 默认路径（可选）
 * @returns 选择的文件路径数组，如果取消则返回 null
 */
export function selectFiles(filters?: DialogFilter[], defaultPath?: string): Promise<string[] | null> {
  return showOpen({
    filters,
    defaultPath,
    multiple: true,
    directory: false
  }) as Promise<string[] | null>;
}

/**
 * 选择文件夹
 * @param defaultPath 默认路径（可选）
 * @returns 选择的文件夹路径，如果取消则返回 null
 */
export function selectFolder(defaultPath?: string): Promise<string | null> {
  return showOpen({
    defaultPath,
    multiple: false,
    directory: true
  }) as Promise<string | null>;
}

/**
 * 保存文件对话框（简化版）
 * @param defaultName 默认文件名（可选）
 * @param filters 文件过滤器（可选）
 * @returns 选择的保存路径，如果取消则返回 null
 */
export function saveFile(defaultName?: string, filters?: DialogFilter[]): Promise<string | null> {
  return showSave({
    defaultPath: defaultName,
    filters
  });
}

// 创建 Dialog API 命名空间
export const dialog = {
  // 核心方法
  showMessage,
  showConfirm,
  showOpen,
  showSave,
  
  // 便捷方法
  info,
  warning,
  error,
  confirm,
  selectFile,
  selectFiles,
  selectFolder,
  saveFile,
  
  // 错误处理工具
  createDialogError,
  isDialogError,
};

File: packages/app/src-tauri/src/plugin_api/dialog.rs
use serde::{Deserialize, Serialize};
use tauri::AppHandle;
use tauri_plugin_dialog::{DialogExt, MessageDialogButtons, MessageDialogKind};
use tokio::sync::oneshot;

#[derive(Debug, Serialize, Deserialize)]
pub struct DialogError {
    pub name: String,
    pub message: String,
    pub code: Option<String>,
}

impl From<String> for DialogError {
    fn from(message: String) -> Self {
        DialogError {
            name: "DialogError".to_string(),
            message,
            code: None,
        }
    }
}

impl From<&str> for DialogError {
    fn from(message: &str) -> Self {
        DialogError {
            name: "DialogError".to_string(),
            message: message.to_string(),
            code: None,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct MessageDialogOptions {
    pub title: Option<String>,
    pub message: String,
    pub kind: Option<String>,
    #[serde(rename = "okLabel")]
    pub ok_label: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ConfirmDialogOptions {
    pub title: Option<String>,
    pub message: String,
    pub kind: Option<String>,
    #[serde(rename = "okLabel")]
    pub ok_label: Option<String>,
    #[serde(rename = "cancelLabel")]
    pub cancel_label: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DialogFilter {
    pub name: String,
    pub extensions: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct OpenDialogOptions {
    pub title: Option<String>,
    #[serde(rename = "defaultPath")]
    pub default_path: Option<String>,
    pub filters: Option<Vec<DialogFilter>>,
    pub multiple: Option<bool>,
    pub directory: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SaveDialogOptions {
    pub title: Option<String>,
    #[serde(rename = "defaultPath")]
    pub default_path: Option<String>,
    pub filters: Option<Vec<DialogFilter>>,
}

// 转换 kind 字符串到 MessageDialogKind
fn parse_message_kind(kind: Option<String>) -> MessageDialogKind {
    match kind.as_deref() {
        Some("info") => MessageDialogKind::Info,
        Some("warning") => MessageDialogKind::Warning,
        Some("error") => MessageDialogKind::Error,
        _ => MessageDialogKind::Info,
    }
}

// 转换 DialogFilter 到文件扩展名数组
fn convert_filters(filters: Option<Vec<DialogFilter>>) -> Vec<(String, Vec<String>)> {
    filters
        .unwrap_or_default()
        .into_iter()
        .map(|f| (f.name, f.extensions))
        .collect()
}

#[tauri::command]
pub async fn plugin_dialog_message(
    app: AppHandle,
    options: MessageDialogOptions,
) -> Result<(), DialogError> {
    let kind = parse_message_kind(options.kind);
    let title = options.title.unwrap_or_else(|| "Message".to_string());

    app.dialog()
        .message(&options.message)
        .title(&title)
        .kind(kind)
        .blocking_show();
    Ok(())
}

#[tauri::command]
pub async fn plugin_dialog_confirm(
    app: AppHandle,
    options: ConfirmDialogOptions,
) -> Result<bool, DialogError> {
    let kind = parse_message_kind(options.kind);
    let title = options.title.unwrap_or_else(|| "Confirm".to_string());

    // 使用异步 API 来创建确认对话框，明确设置为确认/取消按钮
    let (tx, rx) = oneshot::channel();

    app.dialog()
        .message(&options.message)
        .title(&title)
        .kind(kind)
        .buttons(MessageDialogButtons::OkCancel) // 明确设置为确认/取消按钮
        .show(move |confirmed| {
            let _ = tx.send(confirmed);
        });

    // 等待用户选择
    let confirmed = rx.await.unwrap_or(false);
    Ok(confirmed)
}

#[tauri::command]
pub async fn plugin_dialog_open(
    app: AppHandle,
    options: OpenDialogOptions,
) -> Result<Option<serde_json::Value>, DialogError> {
    let mut dialog = app.dialog().file();

    if let Some(title) = options.title {
        dialog = dialog.set_title(&title);
    }

    if let Some(default_path) = options.default_path {
        dialog = dialog.set_directory(&default_path);
    }

    let filters = convert_filters(options.filters);
    for (name, extensions) in filters {
        let ext_refs: Vec<&str> = extensions.iter().map(|s| s.as_str()).collect();
        dialog = dialog.add_filter(&name, &ext_refs);
    }

    // 使用 oneshot channel 来正确处理异步回调
    let (tx, rx) = oneshot::channel();

    if options.directory.unwrap_or(false) {
        // 选择文件夹
        dialog.pick_folder(move |path| {
            let result = path.map(|p| serde_json::Value::String(p.to_string()));
            let _ = tx.send(result);
        });
    } else if options.multiple.unwrap_or(false) {
        // 选择多个文件
        dialog.pick_files(move |paths| {
            let result = paths.map(|paths| {
                let path_strings: Vec<String> = paths.into_iter().map(|p| p.to_string()).collect();
                serde_json::json!(path_strings)
            });
            let _ = tx.send(result);
        });
    } else {
        // 选择单个文件
        dialog.pick_file(move |path| {
            let result = path.map(|p| serde_json::Value::String(p.to_string()));
            let _ = tx.send(result);
        });
    }

    // 等待对话框完成
    let final_result = rx.await.unwrap_or(None);
    Ok(final_result)
}

#[tauri::command]
pub async fn plugin_dialog_save(
    app: AppHandle,
    options: SaveDialogOptions,
) -> Result<Option<String>, DialogError> {
    let mut dialog = app.dialog().file();

    if let Some(title) = options.title {
        dialog = dialog.set_title(&title);
    }

    if let Some(default_path) = options.default_path {
        dialog = dialog.set_file_name(&default_path);
    }

    let filters = convert_filters(options.filters);
    for (name, extensions) in filters {
        let ext_refs: Vec<&str> = extensions.iter().map(|s| s.as_str()).collect();
        dialog = dialog.add_filter(&name, &ext_refs);
    }

    // 使用 oneshot channel 来正确处理异步回调
    let (tx, rx) = oneshot::channel();

    dialog.save_file(move |path| {
        let result = path.map(|p| p.to_string());
        let _ = tx.send(result);
    });

    // 等待对话框完成
    let path_string = rx.await.unwrap_or(None);
    Ok(path_string)
}

